app/javascript/app/service-provider-form.js

```diff
@@ -92,13 +92,83 @@ function protocolOptionSetup() {
   });
 }
 
+function certificateUploadSetup() {
+  // Selectors
+  const fileContainer = document.getElementById('certificate-container');
+  const pemInputMessage = document.querySelector('.js-pem-input-error-message');
+  const pemInput = document.querySelector('.js-pem-input');
+  const pemFilename = document.querySelector('.js-pem-file-name');
+
+  // Functions
+  const setPemError = (message) => {
+    pemInputMessage.textContent = message;
+  };
+
+  const handleUploadedCert = () => {
+    const file = pemInput.files[0];
+
+    pemFilename.textContent = file ? file.name : null;
+
+    if (file && file.text) {
+      file.text().then((content) => {
+        if (content.includes('PRIVATE')) {
+          setPemError('This is a private key, upload the public key instead');
+        } else if (!content.includes('-----BEGIN CERTIFICATE-----')) {
+          setPemError('This file does not appear to be PEM encoded');
+        } else {
+          setPemError(null);
+        }
+      });
+    } else {
+      setPemError(null);
+    }
+  };
+
+  // Event triggers
+  fileContainer.addEventListener('change', handleUploadedCert);
```

Comment:
This should probably be fine due to event bubbling, but looking at how this was implemented previously, and given that we already create a variable for `pemInput`, we could bind the event handler to the input directly and drop the `fileContainer` variable.

```suggestion
  pemInput.addEventListener('change', handleUploadedCert);
```